File: Lighthouse.Backend/Lighthouse.Backend/Models/DeliveryNote.cs
using Lighthouse.Backend.Services.Interfaces;

namespace Lighthouse.Backend.Models
{
    public class DeliveryNote : IEntity
    {
        public int Id { get; set; }

        public int DeliveryId { get; set; }

        public string Text { get; set; } = string.Empty;

        /// <summary>
        /// A genuine instant, so the global <c>Properties&lt;DateTime&gt;()</c> converter is correct
        /// for it. The day a reader sees is reduced from this once, server-side, in the instance's
        /// own zone - the client is never handed an instant to turn into a day itself.
        /// </summary>
        public DateTime CreatedAt { get; set; }

        public DateTime? LastEditedAt { get; set; }

        /// <summary>
        /// Null on an instance with authentication switched off, where there is nobody to name. That
        /// is the value the "may I change this note" rule is written around, so it must stay nullable.
        /// </summary>
        public int? AuthorUserProfileId { get; set; }

        /// <summary>
        /// The author's name as it stood when the note was written. Held separately from the profile
        /// so a durable record does not silently re-label itself when somebody is renamed or removed.
        /// </summary>
        public string? AuthorDisplayName { get; set; }
    }
}

File: Lighthouse.Frontend/src/models/Delivery/DeliveryNote.ts
export interface IDeliveryNote {
	id: number;
	deliveryId: number;
	text: string;
	/** The instant the note was written. Sorting happens on this. */
	createdAt: string;
	/**
	 * The day a reader sees, already reduced on the server in the instance's own zone. Render this
	 * rather than turning `createdAt` into a day here, which lands a day out for readers west of UTC.
	 */
	createdOn: string;
	lastEditedAt?: string | null;
	lastEditedOn?: string | null;
	/** Null when nobody was signed in to name — the note is shown unattributed, never as a placeholder. */
	authorDisplayName?: string | null;
	/**
	 * Whether this reader may correct or withdraw this note. Decided on the server, because the rule
	 * has an awkward case (a note nobody signed) that is not worth having two copies of.
	 */
	canModify?: boolean;
}

export class DeliveryNote implements IDeliveryNote {
	id!: number;
	deliveryId!: number;
	text!: string;
	createdAt!: string;
	createdOn!: string;
	lastEditedAt?: string | null;
	lastEditedOn?: string | null;
	authorDisplayName?: string | null;
	canModify?: boolean;

	static fromBackend(data: IDeliveryNote): DeliveryNote {
		const note = new DeliveryNote();
		note.id = data.id;
		note.deliveryId = data.deliveryId;
		note.text = data.text ?? "";
		note.createdAt = data.createdAt;
		note.createdOn = data.createdOn;
		note.lastEditedAt = data.lastEditedAt ?? null;
		note.lastEditedOn = data.lastEditedOn ?? null;
		note.authorDisplayName = data.authorDisplayName ?? null;
		note.canModify = data.canModify ?? false;
		return note;
	}
}
